### DISCUSSION TITLE
The situation of "multiple learners" is in which code file?

### DISCUSSION THREAD

**Discussion by zonewenke** [opendilab/DI-engine#819] [Q&A] [answered]
Hello, community!
      I came across Key Graphs on the introduction page of the impala algorithm (https://di-engine-docs.readthedocs.io/zh-cn/latest/12_policies/impala.html) and am quite interested in the scenario of multiple learners. However, I couldn't find any code files related to multiple learners in the DI-engine project's codebase.
![微信图片_20240720155027](https://github.com/user-attachments/assets/df09ea9b-84e7-4acc-b3f0-8a621bbb525f)
     I am currently running an experiment with SpaceInvaders (in the new pipeline), and the configuration file is (https://github.com/opendilab/DI-engine/blob/main/dizoo/atari/config/serial/spaceinvaders/spaceinvaders_impala_config.py), but I am not sure whether the configuration in this file is for a single learner or for multiple learners.
    If anyone knows, please let me know, I would be very grateful!

**Comment by PaParaZz1** [comment, marked as answer]
This demo `spaceinvaders_impala_config.py` is a single-worker example. For the multi-learner case, here is a basic [example](https://github.com/opendilab/DI-engine/blob/main/dizoo/atari/config/serial/spaceinvaders/spaceinvaders_dqn_config_multi_gpu_ddp.py) implemented with PyTorch DDP, you can refer to it for more details. If you are interested in other distributed RL cases, you can continue to ask in this discussion.

**Reply by zonewenke** [reply]
Thank you very much for your answer; it's very timely and helpful to me.

**Reply by Backplate** [reply]
Hello, I am also interested in distributed RL, particularly the distributed RL on a cluster.    I looked example you provided above, it used serial_pipeline, and looks like should be launched by torch elastic launch.   This example is quite different compared to the example provided in the Doc distributed section, which shows how to run [DQN script](https://di-engine-docs.readthedocs.io/zh-cn/latest/03_system/distributed.html) on k8s slurm cluster but that one is not a multiple learner case, if I would run the multi-learner case on k8s, is there any reference example could I see?   thank you very much.

**Comment by zonewenke** [comment]
Thank you very much for your timely response, it has been a great help to me!




---Original---
From: ***@***.***&gt;
Date: Mon, Jul 22, 2024 10:54 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [opendilab/DI-engine] The situation of "multiple learners" is inwhich code file? (Discussion #819)




 
This demo spaceinvaders_impala_config.py is a single-worker example. For the multi-learner case, here is a basic example implemented with PyTorch DDP, you can refer to it for more details. If you are interested in other distributed RL cases, you can continue to ask in this discussion.
